internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,435 @@
+		if err != nil {
+			return nil, err
+		}
+		if manager.endpoint, err = parseConnectionString(connectionString, EndpointSection); err != nil {
+			return nil, err
+		}
+		if manager.secret, err = parseConnectionString(connectionString, SecretSection); err != nil {
+			return nil, err
+		}
+		if manager.id, err = parseConnectionString(connectionString, IdSection); err != nil {
+			return nil, err
+		}
+		if staticClient, err = azappconfig.NewClientFromConnectionString(connectionString, clientOptionWithModuleInfo); err != nil {
+			return nil, err
+		}
+	} else {
+		if manager.credential, err = CreateTokenCredential(ctx, provider.Spec.Auth, provider.Namespace); err != nil {
+			return nil, err
+		}
+		if staticClient, err = azappconfig.NewClient(*provider.Spec.Endpoint, manager.credential, clientOptionWithModuleInfo); err != nil {
+			return nil, err
+		}
+		manager.endpoint = *provider.Spec.Endpoint
+	}
+
+	manager.validDomain = getValidDomain(manager.endpoint)
+	manager.StaticClientWrappers = []*ConfigurationClientWrapper{{
+		Endpoint:       manager.endpoint,
+		Client:         staticClient,
+		BackOffEndTime: metav1.Time{},
+		FailedAttempts: 0,
+	}}
+
+	return manager, nil
+}
+
+func (manager *ConfigurationClientManager) GetClients(ctx context.Context) ([]*ConfigurationClientWrapper, error) {
+	currentTime := metav1.Now()
+	clients := make([]*ConfigurationClientWrapper, 0)
+	for _, clientWrapper := range manager.StaticClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	if !manager.ReplicaDiscoveryEnabledEnabled {
+		return clients, nil
+	}
+
+	if currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClientWrappers == nil ||
+			currentTime.After(manager.lastFallbackClientRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		go manager.DiscoverFallbackClients(ctx, url.Host)
+	}
+
+	for _, clientWrapper := range manager.DynamicClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	return clients, nil
+}
+
+func (manager *ConfigurationClientManager) RefreshClients(ctx context.Context) error {
+	currentTime := metav1.Now()
+	if manager.ReplicaDiscoveryEnabledEnabled &&
+		currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.endpoint)
+		go manager.DiscoverFallbackClients(ctx, url.Host)
+	}
+
+	return nil
+}
+
+func (manager *ConfigurationClientManager) DiscoverFallbackClients(ctx context.Context, host string) {
+	srvTargetHosts, err := QuerySrvTargetHost(ctx, host)
+	if err != nil {
+		klog.Warningf("Fail to build fall back clients %s", err.Error())
+	}
```

Comment:
why not return or throw in case of error? It continues to execute below codes when something wrong

Reply:
It should return, updated.